File: src/funcs/appleMusicMeta.js
const https = require("https");

const APPLE_MUSIC_URL_RE = /https:\/\/(?:classical\.)?music\.apple\.com\/([a-z]{2})\/(artist|album|playlist|song|music-video|post)\/[^/]*(?:\/([^/?]*))?(?:\?i=)?([0-9a-z]*)?/;

function parseAppleMusicUrl(url) {
    const match = url.match(APPLE_MUSIC_URL_RE);
    if (!match) return null;

    const storefront = match[1];
    const type = match[2];
    const primaryId = match[3];
    const subId = match[4];

    const id = subId || primaryId;
    const resolvedType = subId ? "song" : type;

    return { storefront, type: resolvedType, id };
}

function itunesLookup(id, country = "us") {
    return new Promise((resolve, reject) => {
        const url = `https://itunes.apple.com/lookup?id=${encodeURIComponent(id)}&country=${encodeURIComponent(country)}&entity=song`;
        const req = https.get(url, { headers: { "User-Agent": "MediaHarbor/1.0" } }, (res) => {
            if ([301, 302, 307, 308].includes(res.statusCode)) {
                https.get(res.headers.location, (res2) => {
                    let data = "";
                    res2.on("data", (chunk) => data += chunk);
                    res2.on("end", () => {
                        try { resolve(JSON.parse(data)); } catch (e) { reject(e); }
                    });
                }).on("error", reject);
                return;
            }
            let data = "";
            res.on("data", (chunk) => data += chunk);
            res.on("end", () => {
                try { resolve(JSON.parse(data)); } catch (e) { reject(e); }
            });
        });
        req.on("error", reject);
        req.setTimeout(10000, () => { req.destroy(); reject(new Error("iTunes API timeout")); });
    });
}

function scaleArtworkUrl(artworkUrl, size = 1200) {
    if (!artworkUrl) return null;
    return artworkUrl.replace(/\d+x\d+(bb|cc|sr)/, `${size}x${size}$1`);
}

async function resolveMetadata(url) {
    try {
        const parsed = parseAppleMusicUrl(url);
        if (!parsed || !parsed.id || !parsed.id.match(/^\d+$/)) {
            return null;
        }

        const data = await itunesLookup(parsed.id, parsed.storefront);
        if (!data || !data.results || data.results.length === 0) return null;

        const results = data.results;

        if (parsed.type === "song") {
            const song = results.find(r => r.wrapperType === "track" && String(r.trackId) === String(parsed.id))
                || results.find(r => r.wrapperType === "track");
            if (song) {
                return {
                    title: song.trackName,
                    artist: song.artistName,
                    album: song.collectionName,
                    coverUrl: scaleArtworkUrl(song.artworkUrl100)
                };
            }
        }

        if (parsed.type === "album") {
            const album = results.find(r => r.wrapperType === "collection")
                || results[0];
            if (album) {
                return {
                    title: album.collectionName,
                    artist: album.artistName,
                    album: album.collectionName,
                    coverUrl: scaleArtworkUrl(album.artworkUrl100)
                };
            }
        }

        const first = results[0];
        if (first) {
            return {
                title: first.trackName || first.collectionName || first.artistName || null,
                artist: first.artistName || null,
                album: first.collectionName || null,
                coverUrl: scaleArtworkUrl(first.artworkUrl100)
            };
        }

        return null;
    } catch (err) {
        return null;
    }
}

module.exports = { parseAppleMusicUrl, resolveMetadata, scaleArtworkUrl };

File: src/core/src/meta/apple_music_meta.rs
use regex::Regex;
use serde::{Deserialize, Serialize};

use crate::{
    errors::MhResult,
    http_client::{build_mozilla_client, fetch_json},
};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AppleMusicUrlInfo {
    pub storefront: String,
    pub content_type: String,
    pub content_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AppleMusicTrackInfo {
    pub title: Option<String>,
    pub artist: Option<String>,
    pub album: Option<String>,
    pub cover_url: Option<String>,
}

pub fn parse_apple_music_url(url: &str) -> Option<AppleMusicUrlInfo> {
    let re = Regex::new(
        r"https://(?:classical\.)?music\.apple\.com/([a-z]{2})/(artist|album|playlist|song|music-video|post)/[^/]*(?:/([^/?]*))?(?:\?i=)?([0-9a-z]*)?"
    ).ok()?;

    let caps = re.captures(url)?;
    let storefront = caps.get(1)?.as_str().to_string();
    let resource_type = caps.get(2)?.as_str().to_string();
    let primary_id = caps.get(3).map(|m| m.as_str()).unwrap_or("");
    let sub_id = caps.get(4).map(|m| m.as_str()).unwrap_or("");

    let (content_id, content_type) = if !sub_id.is_empty() {
        (sub_id.to_string(), "song".to_string())
    } else {
        (primary_id.to_string(), resource_type)
    };

    if content_id.is_empty() {
        return None;
    }

    Some(AppleMusicUrlInfo {
        storefront,
        content_type,
        content_id,
    })
}

pub fn scale_artwork_url(url: &str, size: u32) -> String {
    let re = Regex::new(r"\d+x\d+(bb|cc|sr)").unwrap();
    re.replace(url, format!("{}x{}$1", size, size).as_str())
        .to_string()
}

pub async fn resolve_metadata(url: &str) -> MhResult<Option<AppleMusicTrackInfo>> {
    let parsed = match parse_apple_music_url(url) {
        Some(p) => p,
        None => return Ok(None),
    };

    if !parsed.content_id.chars().all(|c| c.is_ascii_digit()) {
        return Ok(None);
    }

    let client = build_mozilla_client()?;
    let lookup_url = format!(
        "https://itunes.apple.com/lookup?id={}&country={}&entity=song",
        parsed.content_id, parsed.storefront
    );

    let data: serde_json::Value = fetch_json(&client, &lookup_url).await?;
    let results = match data["results"].as_array() {
        Some(r) if !r.is_empty() => r,
        _ => return Ok(None),
    };

    let info = match parsed.content_type.as_str() {
        "song" => {
            let song = results
                .iter()
                .find(|r| {
                    r["wrapperType"].as_str() == Some("track")
                        && r["trackId"].to_string() == parsed.content_id
                })
                .or_else(|| {
                    results
                        .iter()
                        .find(|r| r["wrapperType"].as_str() == Some("track"))
                });
            song.map(|s| AppleMusicTrackInfo {
                title: s["trackName"].as_str().map(str::to_string),
                artist: s["artistName"].as_str().map(str::to_string),
                album: s["collectionName"].as_str().map(str::to_string),
                cover_url: s["artworkUrl100"]
                    .as_str()
                    .map(|u| scale_artwork_url(u, 1200)),
            })
        }
        "album" => {
            let album = results
                .iter()
                .find(|r| r["wrapperType"].as_str() == Some("collection"))
                .or_else(|| results.first());
            album.map(|a| AppleMusicTrackInfo {
                title: a["collectionName"].as_str().map(str::to_string),
                artist: a["artistName"].as_str().map(str::to_string),
                album: a["collectionName"].as_str().map(str::to_string),
                cover_url: a["artworkUrl100"]
                    .as_str()
                    .map(|u| scale_artwork_url(u, 1200)),
            })
        }
        _ => results.first().map(|f| AppleMusicTrackInfo {
            title: f["trackName"]
                .as_str()
                .or_else(|| f["collectionName"].as_str())
                .or_else(|| f["artistName"].as_str())
                .map(str::to_string),
            artist: f["artistName"].as_str().map(str::to_string),
            album: f["collectionName"].as_str().map(str::to_string),
            cover_url: f["artworkUrl100"]
                .as_str()
                .map(|u| scale_artwork_url(u, 1200)),
        }),
    };

    Ok(info)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_song_url() {
        let url = "https://music.apple.com/us/album/bohemian-rhapsody/1440806041?i=1440806768";
        let info = parse_apple_music_url(url).unwrap();
        assert_eq!(info.storefront, "us");
        assert_eq!(info.content_type, "song");
        assert_eq!(info.content_id, "1440806768");
    }

    #[test]
    fn scale_url() {
        let url = "https://is1-ssl.mzstatic.com/image/thumb/Music/a.jpg/100x100bb.jpg";
        let scaled = scale_artwork_url(url, 1200);
        assert!(scaled.contains("1200x1200bb"));
    }
}
